Use multiple hosts in restclient

dshulyak opened the pull request:
The goal of this change is not to provide load balancing, but rely on multiple masters in clients and remove single point of failure


<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30588)

<!-- Reviewable:end -->

luxas commented on cluster/addons/dns/kube2sky/kube2sky.go:
```diff
@@ -531,7 +531,7 @@ func newKubeClient() (*kclient.Client, error) {
 	if masterURL != "" && *argKubecfgFile == "" {
 		// Only --kube-master-url was provided.
 		config = &restclient.Config{
-			Host:          masterURL,
```
Hmm, I think kube2sky is deprecated and shouldn't be used anymore... Probably not worth to add effort into that

dshulyak commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -154,8 +154,8 @@ func (ms *MinionServer) launchProxyServer() {
-	if ms.clientConfig.Host != "" {
-		args = append(args, fmt.Sprintf("--master=%s", ms.clientConfig.Host))
+	if len(ms.clientConfig.Hosts) > 0 {
+		args = append(args, fmt.Sprintf("--master=%s", strings.Join(ms.clientConfig.Hosts, ",")))
```
change to api-servers

dshulyak commented on pkg/client/restclient/client.go:
```diff
@@ -188,9 +190,9 @@ func (c *RESTClient) Verb(verb string) *Request {
 
 	if c.Client == nil {
-		return NewRequest(nil, verb, c.base, c.versionedAPIPath, c.contentConfig, c.serializers, backoff, c.Throttle)
+		return NewRequest(nil, verb, c.urlProvider, c.versionedAPIPath, c.contentConfig, c.serializers, backoff, c.Throttle)
```
clarify the point of this, might be removed

dshulyak commented:
```
FAIL	k8s.io/kubernetes/plugin/pkg/scheduler/algorithmprovider/defaults	1.180s
ok  	k8s.io/kubernetes/plugin/pkg/scheduler/api/validation	1.146s
FAIL	k8s.io/kubernetes/plugin/pkg/scheduler/factory	1.264s
```
I don't have this errors locally, what could be the reason for them?

smarterclayton commented on pkg/client/restclient/client.go:
```diff
@@ -46,7 +46,7 @@ const (
 // Most consumers should use client.New() to get a Kubernetes API client.
 type RESTClient struct {
-	urlProvider func() *url.URL
+	urlProvider *URLProvider
```
If you're going to abstract this, make it an interface.

luxas replied:
> EDIT: Should just be an interface, not a pointer to an interface (urlProvider URLProvider)

ping @dshulyak (since editing comments doesn't create notifications)

smarterclayton commented on pkg/client/restclient/request.go:
```diff
@@ -817,7 +819,11 @@ func (r *Request) request(fn func(*http.Request, *http.Response)) error {
-			return err
+			if r.urlProvider.Next() == initialUrl {
+				return err
+			} else {
```
style: if you return/continue in an if, you should never have an else.

smarterclayton commented on pkg/client/restclient/config.go:
```diff
@@ -45,11 +45,11 @@ const (
+	// apiservers. If a URL is given then the (optional) Path of that URL represents a prefix
+	// that must be appended to all request URIs used to access the apiserver. This allows a
+	// frontend proxy to easily relocate all of the apiserver endpoints.
+	Hosts []string
```
Rather than change this, I would prefer:

```
Host string
AlternateHosts []string
```

That reduces the scope of your PR here significantly and confines alternate hosts to the use case it was intended for.  A user must always specify at least one host, but alternates are checked afterwards.

justinsb replied:
+1  - I think this is considered part of our API, also, so we should really avoid changes?

dshulyak replied:
@smarterclayton i am not sure that it will affect amount of lines changed, i will still have to pass those AlternateHosts all over the place, but i will check it. What would be the other reason to change this? I think that it only introduces ambiguity, all those api-servers are equal and which one i, as an operator, should select, what it will affect

@justinsb you mean that this restclient can be used by external apps?

mikedanese replied:
I don't know how we could add this feature without making changes to the Kubeconfig. I'm hoping to consolidate the fields once we implement https://github.com/kubernetes/kubernetes/issues/30395

smarterclayton replied:
> all those api-servers are equal and which one i, as an operator, should select, what it will affect

Even with this PR that isn't the case - the operator may very well know better than you which one they want to use primarily.  I do think some of the changes below are cleaner, but this is a part of our API and will have a ripple impact on our clients.

I think we will have to make changes to kubeconfig, but `alternateHosts` would at least be additive, and old kubeconfigs would continue to work.  If we end ups with `alternateHosts` in the kubeconfig, then we might as well have them internally.

dshulyak replied:
to summarize:
1. in this change --master and --api-servers will overwrite values provided in kubeconfig Server/Servers fields, not clear if i should add --master flag to kubelet
2. Server will be passed as Host, Servers as AlternateHosts
3. i will add Host back to restclient.Config which, if not empty and not present in AlternateHosts, will be used as a first url passed to url provider, this indeed will reduce amount of changes, especially in tests

mikedanese commented:
@dshulyak Thanks for finishing this up! cc @kubernetes/sig-api-machinery @krousey

justinsb commented on cmd/kube-proxy/app/options/options.go:
```diff
@@ -64,6 +65,7 @@ func NewProxyConfig() *ProxyServerConfig {
 func (s *ProxyServerConfig) AddFlags(fs *pflag.FlagSet) {
 	fs.Var(componentconfig.IPVar{Val: &s.BindAddress}, "bind-address", "The IP address for the proxy server to serve on (set to 0.0.0.0 for all interfaces)")
 	fs.StringVar(&s.Master, "master", s.Master, "The address of the Kubernetes API server (overrides any value in kubeconfig)")
+	fs.StringSliceVar(&s.APIServerList, "api-servers", []string{}, "List of Kubernetes API servers for publishing events, and reading pods and services. (ip:port), comma separated.")
```
Not always an IP.

mattymo replied:
Good point. Should be (host:port)

smarterclayton replied:
Port is optional.

justinsb commented on pkg/client/restclient/request.go:
```diff
@@ -691,8 +691,8 @@ func (r *Request) Watch() (watch.Interface, error) {
-	if req.baseURL != nil {
-		url = req.baseURL.Host
+	if req.urlProvider.Get() != nil {
+		url = req.urlProvider.Get().Host
```
Is there any guarantee Get() returns the same both times?

justinsb commented on pkg/client/restclient/urlprovider.go:
```diff
@@ -0,0 +1,54 @@
+)
+
+type URLProvider interface {
+	Get() *url.URL
```
Comment needed to explain the contract here

justinsb commented on pkg/client/restclient/urlprovider.go:
```diff
@@ -0,0 +1,54 @@
+	} else {
+		p.current++
+	}
+	return p.Get()
```
Can you just inline this, and thus drop to a standard mutex?

justinsb commented on pkg/client/restclient/urlprovider.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+type RoundRobinProvider struct {
+	sync.RWMutex
```
Should make clear what the mutex is guarding.

The pattern I've seen is:

```
constants

mutex
stuff guarded by mutex
```

justinsb commented on pkg/client/restclient/urlprovider.go:
```diff
@@ -0,0 +1,54 @@
+	return p.urls[p.current]
+}
+
+func (p *RoundRobinProvider) Next() *url.URL {
```
I think we need a comment somewhere about what health checking and bad node avoidance this does and does not do, and the requirements on the caller to make that happen.

smarterclayton commented on pkg/client/restclient/config.go:
```diff
@@ -112,6 +113,9 @@ type Config struct {
 	// Do we need this?
 	// Version string
+
+	// Hosts must be a slice of strings, with properties of Host
```
Add a bit more Godoc about how this is intended to be used.

smarterclayton replied:
Also, the Godoc type is wrong

smarterclayton commented on pkg/client/restclient/config.go:
```diff
@@ -152,6 +156,22 @@ type ContentConfig struct {
 }
 
+// Returns Host and AlternateHosts as a single slice, without duplicates
+func (c *Config) AllHosts() []string {
```
Could just be "Hosts()"

smarterclayton commented on pkg/client/restclient/config.go:
```diff
@@ -152,6 +156,22 @@ type ContentConfig struct {
+// Returns Host and AlternateHosts as a single slice, without duplicates
+func (c *Config) AllHosts() []string {
+	var hosts []string
+	if c.Host != "" {
```
nit: by convention in our codebase we try to use `len(c.Host) > 0`

smarterclayton commented on pkg/client/restclient/config.go:
```diff
@@ -283,11 +302,16 @@ func InClusterConfig() (*Config, error) {
 func IsConfigTransportTLS(config Config) bool {
-	baseURL, _, err := defaultServerUrlFor(&config)
-	if err != nil {
+	hosts, _, err := defaultServerUrlFor(&config)
```
I think `defaultServerUrlFor` should return `err != nil` if hosts would be empty.

smarterclayton commented on pkg/client/restclient/client.go:
```diff
@@ -103,8 +104,9 @@ func NewRESTClient(baseURL *url.URL, versionedAPIPath string, config ContentConf
+	glog.V(4).Infof("Rest client initialized with following urls %v", hosts)
 	return &RESTClient{
-		base:             &base,
+		urlProvider:      &RoundRobinProvider{urls: hosts},
```
I'd be slightly in favor of RoundRobinProvider implement URLProvider by value, rather than with a pointer.

dshulyak replied:
and use pointer when Request object will be created? But then on each new request we will try to connect to potentially broken api-server

smarterclayton commented on pkg/client/restclient/request.go:
```diff
@@ -660,11 +665,11 @@ func (r *Request) Watch() (watch.Interface, error) {
 	resp, err := client.Do(req)
 	updateURLMetrics(r, resp, err)
-	if r.baseURL != nil {
+	if r.urlProvider.Get() != nil {
```
Would be better if you did

```
if url := r.urlProvider.Get(); url != nil {
  if err != nil {
    r.backoffMgr.UpdateBackoff(url, err, 0)
  ...
```

smarterclayton commented on pkg/client/restclient/request.go:
```diff
@@ -728,14 +733,15 @@ func (r *Request) Stream() (io.ReadCloser, error) {
 	resp, err := client.Do(req)
 	updateURLMetrics(r, resp, err)
-	if r.baseURL != nil {
+	if r.urlProvider.Get() != nil {
```
I don't like calling r.urlProvider.Get() as a way to know whether r.URL() would return nil.  I would prefer this to be the behavior:

```
if url := r.URL(); url == nil {
   if err != nil {
     r.backoffMgr.UpdateBackoff(url, ...)
     ...
```

It's not obvious to a casual browser that r.urlProvider.Get() and r.URL() are coupled - one should dominate / hide the details of the other.

smarterclayton commented on pkg/client/restclient/request.go:
```diff
@@ -800,6 +806,7 @@ func (r *Request) request(fn func(*http.Request, *http.Response)) error {
 	retries := 0
+	initialUrl := r.urlProvider.Get()
 	for {
 		url := r.URL().String()
```
Are we calling `r.URL()` here multiple times and expecting the same value?  If so, we should really stop doing that and return it once.

smarterclayton commented on pkg/kubectl/cmd/clusterinfo.go:
```diff
@@ -60,6 +61,9 @@ func RunClusterInfo(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command) error
 	}
 	printService(out, "Kubernetes master", client.Host)
+	if len(client.AlternateHosts) > 0 {
+		printService(out, "Kubernetes other api servers", strings.Join(client.AlternateHosts, ","))
```
`Kubernetes alternate masters`

smarterclayton commented on pkg/client/restclient/request_test.go:
```diff
@@ -1543,7 +1541,7 @@ func testRESTClient(t testing.TB, srv *httptest.Server) *RESTClient {
 	}
 	versionedAPIPath := testapi.Default.ResourcePath("", "", "")
-	client, err := NewRESTClient(baseURL, versionedAPIPath, defaultContentConfig(), 0, 0, nil, nil)
+	client, err := NewRESTClient([]*url.URL{baseURL}, versionedAPIPath, defaultContentConfig(), 0, 0, nil, nil)
```
Please add a new test to request_test.go that verifies that retries properly walk through other alternate hosts and terminate on errors, etc (we really need 100% coverage of request.go, since it's the core of the whole project)

dshulyak replied:
added

krousey commented:
Reviewed 5 of 51 files at r1, 2 of 16 files at r2, 4 of 32 files at r3, 1 of 2 files at r4, 1 of 5 files at r6, 19 of 62 files at r7.
Review status: all files reviewed at latest revision, 31 unresolved discussions, some commit checks failed.

---

*[pkg/client/restclient/request.go, line 132 \[r7\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588#-KPOMslWe3_MKOt1Knym:-KPOMslWe3_MKOt1Knyn:b-swd421) ([raw file](https://github.com/kubernetes/kubernetes/blob/5d23bbbb27e84272612d7406ba52963dfed88540/pkg/client/restclient/request.go#L132)):*
> ```Go
> 		backoff = &NoBackoff{}
> 	}
> 	baseUrl := urlProvider.Get()
> ```

Style nit: this should be baseURL.

---

*[pkg/client/restclient/request.go, line 268 \[r7\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588#-KPOMydpiUk79SwENOCo:-KPOMydpiUk79SwENOCp:bu1b0hl) ([raw file](https://github.com/kubernetes/kubernetes/blob/5d23bbbb27e84272612d7406ba52963dfed88540/pkg/client/restclient/request.go#L268)):*
> ```Go
> 		return r
> 	}
> 	baseUrl := r.urlProvider.Get()
> ```

baseURL

---

*[pkg/client/restclient/urlprovider.go, line 49 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588#-KP8Clqvt2DkeqMgMd7F-r5-49:-KPOPCrziTVRgZb5_CSY:b-fhlbw0) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c00bb0567a366ac91a5fef36647c64fbd619d3f/pkg/client/restclient/urlprovider.go#L49)):*
<details><summary><i>Previously, justinsb (Justin Santa Barbara) wrote…</i></summary>
> Can you just inline this, and thus drop to a standard mutex?
</details>
I think he meant just `return p.urls[p.current]` instead of calling `p.Get()` and stick with the normal lock/defer unlock pattern.

---

*[pkg/client/restclient/urlprovider.go, line 24 \[r7\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588#-KPOOHEMhxDD1dYrcDI1:-KPOOHEN-Ybs-1btlWiL:b-t9kbcs) ([raw file](https://github.com/kubernetes/kubernetes/blob/5d23bbbb27e84272612d7406ba52963dfed88540/pkg/client/restclient/urlprovider.go#L24)):*
> ```Go
> )
> 
> type URLProvider interface {
> ```

This type needs a doc string

---

*[pkg/client/restclient/urlprovider.go, line 32 \[r7\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588#-KPOOLILHqOu4mWFd_dI:-KPOOLILHqOu4mWFd_dJ:b-pdu6uw) ([raw file](https://github.com/kubernetes/kubernetes/blob/5d23bbbb27e84272612d7406ba52963dfed88540/pkg/client/restclient/urlprovider.go#L32)):*
> ```Go
> }
> 
> type RoundRobinProvider struct {
> ```

This type also needs a doc string. This is probably where the health checking info that @justinb wants should go.

---

*[pkg/client/restclient/urlprovider.go, line 41 \[r7\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588#-KPOP0jsH8QLc4-3hrLw:-KPOP0jsH8QLc4-3hrLx:b-az40ek) ([raw file](https://github.com/kubernetes/kubernetes/blob/5d23bbbb27e84272612d7406ba52963dfed88540/pkg/client/restclient/urlprovider.go#L41)):*
> ```Go
> 	p.Lock()
> 	defer p.Unlock()
> 	if len(p.urls)-1 < p.current {
> ```

What is this guarding against? an empty slice? I would consider it an error to construct a `RoundRobinProvider` with an empty slice.

---

*[pkg/client/restclient/urlprovider.go, line 59 \[r7\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588#-KPOOYd49hbmTo_pQB-d:-KPOOYd49hbmTo_pQB-e:b-9ztcrh) ([raw file](https://github.com/kubernetes/kubernetes/blob/5d23bbbb27e84272612d7406ba52963dfed88540/pkg/client/restclient/urlprovider.go#L59)):*
> ```Go
> }
> 
> func NewRoundRobinProvider(urls ...*url.URL) *RoundRobinProvider {
> ```

I know it's tedious, but please provide doc strings for all public types and functions.

---

*[pkg/client/restclient/urlprovider_test.go, line 28 \[r7\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588#-KPOPX_iwrqcnp159LyJ:-KPOPX_j8KwM3LT1Ys-B:b-eq2ays) ([raw file](https://github.com/kubernetes/kubernetes/blob/5d23bbbb27e84272612d7406ba52963dfed88540/pkg/client/restclient/urlprovider_test.go#L28)):*
> ```Go
> 	url2, _ := url.Parse("http://master:8001")
> 
> 	rr := &RoundRobinProvider{urls: []*url.URL{url1, url2}}
> ```

Why not call your new function here?

---

*[pkg/controller/replicaset/replica_set_test.go, line 1077 \[r7\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588#-KPOQbJPY_i3xffOSWKw:-KPOQbJPY_i3xffOSWKx:b-8o6k9j) ([raw file](https://github.com/kubernetes/kubernetes/blob/5d23bbbb27e84272612d7406ba52963dfed88540/pkg/controller/replicaset/replica_set_test.go#L1077)):*
> ```Go
> // and with garbageCollectorEnabled set to true
> func setupManagerWithGCEnabled() (manager *ReplicaSetController, fakePodControl *controller.FakePodControl) {
> 	c := clientset.NewForConfigOrDie(&restclient.Config{Host: "0.0.0.0:8000", ContentConfig: restclient.ContentConfig{GroupVersion: testapi.Default.GroupVersion()}})
> ```

Why are the rest removed, but this one is given a value?

---

*[pkg/controller/replication/replication_controller_test.go, line 1140 \[r7\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588#-KPOQiQQO0iY9iPwc6mv:-KPOQiQQO0iY9iPwc6mw:bnh7et3) ([raw file](https://github.com/kubernetes/kubernetes/blob/5d23bbbb27e84272612d7406ba52963dfed88540/pkg/controller/replication/replication_controller_test.go#L1140)):*
> ```Go
> // setupManagerWithGCEnabled creates a RC manager with a fakePodControl and with garbageCollectorEnabled set to true
> func setupManagerWithGCEnabled() (manager *ReplicationManager, fakePodControl *controller.FakePodControl) {
> 	c := clientset.NewForConfigOrDie(&restclient.Config{Host: "0.0.0.0:8000", ContentConfig: restclient.ContentConfig{GroupVersion: testapi.Default.GroupVersion()}})
> ```

Why remove the rest but give this one a value?

---

*[test/e2e/framework/framework.go, line 136 \[r7\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588#-KPOR6ULTDs_K6yZGUMi:-KPOR6ULTDs_K6yZGUMj:ban7u9r) ([raw file](https://github.com/kubernetes/kubernetes/blob/5d23bbbb27e84272612d7406ba52963dfed88540/test/e2e/framework/framework.go#L136)):*
> ```Go
> // configs may share some common struct.
> func getClientRepoConfig(src *restclient.Config) (dst *clientreporestclient.Config) {
> 	skippedFields := sets.NewString("Transport", "WrapTransport", "RateLimiter", "AuthConfigPersister", "AlternateHosts")
> ```

@caesarxuchao is this to do with your client repo?

---

*[test/e2e_node/image_conformance_test.go, line 34 \[r7\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588#-KPORGvxH4QoVKuQJUj9:-KPORGvxH4QoVKuQJUjA:b-hxzd4e) ([raw file](https://github.com/kubernetes/kubernetes/blob/5d23bbbb27e84272612d7406ba52963dfed88540/test/e2e_node/image_conformance_test.go#L34)):*
> ```Go
> 
> var _ = Describe("Image Container Conformance Test", func() {
> 	Describe("image conformance blackbox test", func() {
> ```

Why is this test no longer being marked flaky? Is this PR supposed to fix a flake here?

---

*[test/e2e_node/util.go, line 29 \[r7\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588#-KPORxfw2Gok-lEKOQYB:-KPORxfxDj4nibxyDCLh:bvcr4zt) ([raw file](https://github.com/kubernetes/kubernetes/blob/5d23bbbb27e84272612d7406ba52963dfed88540/test/e2e_node/util.go#L29)):*
> ```Go
> var stopServices = flag.Bool("stop-services", true, "If true, stop local node services after running tests")
> 
> type SharedContext struct {
> ```

Why is this being added?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588)*
<!-- Sent from Reviewable.io -->

krousey commented:
Reviewed 1 of 62 files at r7, 14 of 16 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 33 unresolved discussions.

---

*[pkg/client/restclient/request.go, line 133 \[r10\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588#-KPUW-J9ydqMtx-KWtFO:-KPUW-J9ydqMtx-KWtFP:b-ub2ta8) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c00346fe6e7b3652dcf49a977a5176e1c140754/pkg/client/restclient/request.go#L133)):*
> ```Go
> 	}
> 	pathPrefix := "/"
> 	if url := urlProvider.Get(); url != nil {
> ```

You should use `u` instead of `url`, not because `url` isn't short enough, but because it's a package name that you're shadowing.

---

*[pkg/client/restclient/request.go, line 590 \[r10\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588#-KPUWIIqzK7VYeuSjqjo:-KPUWIIqzK7VYeuSjqjp:bdhpx4j) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c00346fe6e7b3652dcf49a977a5176e1c140754/pkg/client/restclient/request.go#L590)):*
> ```Go
> 
> 	finalURL := &url.URL{}
> 	if url := r.urlProvider.Get(); url != nil {
> ```

Same comment about shadowing the package name.

---

*[pkg/client/restclient/urlprovider.go, line 24 \[r7\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588#-KPOOHEMhxDD1dYrcDI1:-KPUdPH_dX5OGmu3JgIl:bnc15yr) ([raw file](https://github.com/kubernetes/kubernetes/blob/5d23bbbb27e84272612d7406ba52963dfed88540/pkg/client/restclient/urlprovider.go#L24)):*
<details><summary><i>Previously, krousey wrote…</i></summary>
> This type needs a doc string
</details>
In general, you should use full sentences and also use the name of what you're documenting. E.g.


https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588)*
<!-- Sent from Reviewable.io -->

dshulyak commented:
@krousey i should squash commits into single, right?

smarterclayton commented on cmd/kubelet/app/options/options.go:
```diff
@@ -75,7 +75,7 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.Var(&s.AuthPath, "auth-path", "Path to .kubernetes_auth file, specifying how to authenticate to API server.")
 	fs.MarkDeprecated("auth-path", "will be removed in a future version")
-	fs.StringSliceVar(&s.APIServerList, "api-servers", []string{}, "List of Kubernetes API servers for publishing events, and reading pods and services. (ip:port), comma separated.")
+	fs.StringSliceVar(&s.APIServerList, "api-servers", []string{}, "List of Kubernetes API servers for publishing events, and reading pods and services. (host:port (port is optional)), comma separated.")
```
Can drop the parenthesis:

`List of Kubernetes API servers for publishing events and reading pods and services. Comma separated list of hostnames or host:port pairs.`

smarterclayton commented on pkg/client/restclient/client.go:
```diff
@@ -78,13 +78,14 @@ type Serializers struct {
-		base.Path += "/"
+func NewRESTClient(hosts []*url.URL, versionedAPIPath string, config ContentConfig, maxQPS float32, maxBurst int, rateLimiter flowcontrol.RateLimiter, client *http.Client) (*RESTClient, error) {
+	for _, host := range hosts {
+		if !strings.HasSuffix(host.Path, "/") {
```
This mutates the input URL, which the previous code did not do.  You'll still need to copy host using *host.

smarterclayton commented on pkg/client/restclient/config.go:
```diff
@@ -152,6 +157,22 @@ type ContentConfig struct {
 	NegotiatedSerializer runtime.NegotiatedSerializer
 }
 
+// Returns Host and AlternateHosts as a single slice, without duplicates
```
Comment is not quite correct, does not prevent duplicates in AlternateHosts, only prevents Host from appearing twice.

smarterclayton commented on pkg/client/restclient/config.go:
```diff
@@ -283,11 +303,16 @@ func InClusterConfig() (*Config, error) {
 // Note: the Insecure flag is ignored when testing for this value, so MITM attacks are
 // still possible.
```
Add to the comment that *any* non HTTPs host will return false.  

Also, what are the security implications of this?  I don't know that we should allow an admin to mix security types - I think we should return an error much earlier in the config process that guards against that (must be all http or all https)

dshulyak replied:
i will add validation to clientcmd/client_config.go

smarterclayton commented on pkg/client/restclient/request_test.go:
```diff
@@ -885,19 +884,15 @@ func TestRequestDo(t *testing.T) {
 		Err     bool
 	}{
 		{
-			Request: &Request{err: errors.New("bail")},
```
Is this test case not represented anywhere?  A predetermined error should still exit out.

smarterclayton commented on pkg/client/restclient/request_test.go:
```diff
@@ -915,6 +910,73 @@ func TestRequestDo(t *testing.T) {
+		Expected map[string]int
+	}{
+		{Request: &Request{
+			urlProvider: NewRoundRobinProvider(firstURL, secondURL),
```
Add a test condition for both URLs are the same, only one URL, and three URLs.

dshulyak replied:
added

justinblayney commented:
please remove me from email updates, i never signed up for them, the mute thread button does not work

smarterclayton commented on pkg/client/restclient/urlprovider_test.go:
```diff
@@ -0,0 +1,42 @@
+	"testing"
+)
+
+func TestRoundRobinProvider(t *testing.T) {
```
Add a test for a single arg and two of the same.

dshulyak replied:
added

smarterclayton commented on pkg/client/unversioned/clientcmd/api/v1/types.go:
```diff
@@ -58,6 +58,8 @@ type Preferences struct {
 type Cluster struct {
 	// Server is the address of the kubernetes cluster (https://hostname:port).
 	Server string `json:"server"`
+	// Servers is an array of addresses to the kubernetes apiservers with the format: (https://hostname:port)
```
In the Godoc describe how this is used (joined with Server, de-duped)

smarterclayton replied:
Also mention in the `Server` Godoc that this is joined with Servers

smarterclayton commented on pkg/client/unversioned/clientcmd/client_config.go:
```diff
@@ -98,12 +97,12 @@ func (config *DirectClientConfig) ClientConfig() (*restclient.Config, error) {
-		clientConfig.Host = u.String()
+
+	clientConfig.AlternateHosts = configClusterInfo.Servers
+	if configClusterInfo.Server != "" {
```
Hrm - this is not the same behavior.  Did you move this stripping of query and fragment somewhere else?

dshulyak replied:
sorry missed this change, will add it back

smarterclayton commented on pkg/kubectl/cmd/clusterinfo.go:
```diff
@@ -60,6 +61,9 @@ func RunClusterInfo(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command) error
 		return err
 	}
 	printService(out, "Kubernetes master", client.Host)
+	if len(client.AlternateHosts) > 0 {
```
Since you added `Servers` to kubeconfig, don't forget to add it to `kubectl config view`

dshulyak replied:
you mean that AlternateHosts should be merged with Servers when user executes `kubectl config view`?

dshulyak replied:
Sorry, i can't figure out what i suppose to change. The only related option in kubectl is:
`-s, --server='': The address and port of the Kubernetes API server` and it doesnt affect `kubectl config view` in any way. Beside that this is how cluster part of my config looks like:

```
clusters:
- cluster:
    certificate-authority-data: REDACTED
    server: https://10.245.1.2
    servers:
    - https://10.245.1.3
    - https://10.245.1.4
  name: vagrant```

smarterclayton replied:
Does not have to be merged, just also displayed at minimum.


On Aug 22, 2016, at 8:59 AM, Dmitry Shulyak <notifications@github.com>
wrote:

> @@ -60,6 +61,9 @@ func RunClusterInfo(f *cmdutil.Factory, out io.Writer,
cmd *cobra.Command) error

>  		return err
>  	}
>  	printService(out, "Kubernetes master", client.Host)
> +	if len(client.AlternateHosts) > 0 {

you mean that AlternateHosts should be merged with Servers when user
executes kubectl config view?

smarterclayton commented:
@mikedanese since this is a subset of the behavior w.r.t. retry from your other PR, can you comment on what retry scenarios + failure detection you think is absolutely critical for 1.4, vs things we can add later?

justinsb commented on pkg/client/restclient/client_test.go:
```diff
@@ -49,7 +49,6 @@ type TestParam struct {
 func TestDoRequestSuccess(t *testing.T) {
 	testServer, fakeHandler, status := testServerEnv(t, 200)
 	defer testServer.Close()
-
```
Please remove this change - it'll just cause merge conflicts for you or others.

justinsb commented on pkg/client/restclient/config.go:
```diff
@@ -278,16 +299,29 @@ func InClusterConfig() (*Config, error) {
+func IsConfigTransportTLS(config Config) (bool, error) {
+	hosts, _, err := defaultServerUrlFor(&config)
+	allHttps := 0
+	// TODO Propagate errors from defaultServerUrlFor
```
Indeed.  At least warn about it.  But why not just return `false, err` (maybe with wrapping)?

dshulyak replied:
in some version i changed it to propagate errors but got several failures in test, i would prefer to sort this out in separate pr if you don't mind

justinsb replied:
What is the problem in those tests?  Maybe add a comment to the TODO explaining the problem if that is the case.

dshulyak replied:
that was existing behaviour, i tried to change it and several tests failed. can i work on it in separate change?

justinsb commented on pkg/client/restclient/request.go:
```diff
@@ -120,20 +120,19 @@ type Request struct {
 	pathPrefix := "/"
-	if baseURL != nil {
-		pathPrefix = path.Join(pathPrefix, baseURL.Path)
+	if u := urlProvider.Get(); u != nil {
```
Ah - does this mean this is in fact _not_ a URLProvider, but a HostProvider?  If we can't mix Path values, we probably shouldn't be allowing it in the contract.

I think the answer is likely to push pathPrefix down so we resolve it whenever we call urlprovider.Get

dshulyak replied:
thanks, looks like simple change in URL method

justinsb commented on pkg/client/restclient/url_utils.go:
```diff
@@ -27,27 +27,46 @@ import (
+
+}
+
+func DefaultServerURLsForHosts(hosts []string, apiPath string, groupVersion unversioned.GroupVersion, defaultTLS bool) ([]*url.URL, string, error) {
```
Needs a comment.  In particular to support moving the localhost default into this function.

dshulyak replied:
returning localhost if empty is a previous behaviour

justinsb replied:
This function is new, and exported.  It should have a comment regardless.

But you also need to support moving the empty-slice logic from defaultServerUrlFor into this function, likely by explaining the purpose of DefaultServerURLsForHosts

justinsb commented on pkg/client/restclient/urlprovider.go:
```diff
@@ -0,0 +1,69 @@
+
+// A URLProvider allows to use multiple URLs.
+type URLProvider interface {
+	// Get returns currently selected URL.
```
Maybe put a sentence about the notion of the "current selected URL" into the comment for URLProvider itself, a high level introduction to the abstraction.

dshulyak replied:
added

justinsb commented on pkg/client/restclient/urlprovider.go:
```diff
@@ -0,0 +1,69 @@
+	// health checking/load balancing, it might be as simple as an iterator.
+	Next() *url.URL
+	// Total returns number of available URLs.
+	Total() int
```
I don't think you want this in your URLProvider contract

dshulyak replied:
maybe i shouldn't be adding this interface at all and just use RoundRobinProvider everywhere

justinsb commented on pkg/client/restclient/url_utils.go:
```diff
@@ -64,30 +83,23 @@ func DefaultServerURL(host, apiPath string, groupVersion unversioned.GroupVersio
 // defaultServerUrlFor is shared between IsConfigTransportTLS and RESTClientFor. It
 // requires Host and Version to be set prior to being called.
-func defaultServerUrlFor(config *Config) (*url.URL, string, error) {
+func defaultServerUrlFor(config *Config) ([]*url.URL, string, error) {
```
Presumably should not be pluralized - defaultServerUrlsFor

justinsb commented on pkg/client/restclient/versions.go:
```diff
@@ -46,10 +46,11 @@ func ServerAPIVersions(c *Config) (groupVersions []string, err error) {
 	if err != nil {
 		return nil, err
 	}
+	baseURL := baseURLs[0]
```
When is version called?  We need a comment to justify that taking the first one is safe:

* If the first server is down
* If the servers have different versions

justinsb commented on pkg/client/unversioned/clientcmd/client_config.go:
```diff
@@ -88,6 +88,21 @@ func (config *DirectClientConfig) RawConfig() (clientcmdapi.Config, error) {
+	if err != nil {
+		return "", err
+	}
+	if u.Opaque != "" && len(u.Path) <= 1 {
```
Can you add a comment here (or somewhere around this function) with some example values that you are trying to filter out

dshulyak replied:
i am not the author of this filter, just moved it to function, so that i can reuse it for collection

justinsb commented on pkg/controller/daemon/daemoncontroller_test.go:
```diff
@@ -135,7 +135,7 @@ func addPods(podStore cache.Store, nodeName string, label map[string]string, num
 }
 
 func newTestController() (*DaemonSetsController, *controller.FakePodControl) {
-	clientset := clientset.NewForConfigOrDie(&restclient.Config{Host: "", ContentConfig: restclient.ContentConfig{GroupVersion: testapi.Default.GroupVersion()}})
```
Do we have to remove the `Host: ""` lines?

justinblayney replied:
Please stop spamming me with email updates about this, The mute link does not work. Someone incorrectly mentioned me in some comment now I get 20-30 email spam messages a day from this thread. Someone remove me.

justinsb commented on test/e2e_node/image_conformance_test.go:
```diff
@@ -32,6 +32,7 @@ const (
 
 var _ = Describe("Image Container Conformance Test", func() {
 	Describe("[Flaky] image conformance blackbox test", func() {
+
```
Remove from PR

justinblayney replied:
Please stop spamming me with email updates about this, The mute link does not work. Someone incorrectly mentioned me in some comment now I get 20-30 email spam messages a day from this thread. Someone remove me.

justinsb commented on test/e2e_node/runtime_conformance_test.go:
```diff
@@ -38,7 +38,6 @@ const (
 
 var _ = framework.KubeDescribe("Container Runtime Conformance Test", func() {
 	f := framework.NewDefaultFramework("runtime-conformance")
-
```
Remove from PR

justinsb commented on test/integration/pods/pods_test.go:
```diff
@@ -37,7 +37,9 @@ func TestPodUpdateActiveDeadlineSeconds(t *testing.T) {
 	ns := framework.CreateTestingNamespace("pod-activedeadline-update", s, t)
 	defer framework.DeleteTestingNamespace(ns, s, t)
 
-	client := client.NewOrDie(&restclient.Config{Host: s.URL, ContentConfig: restclient.ContentConfig{GroupVersion: testapi.Default.GroupVersion()}})
```
Just formatting?  Please remove if so

justinblayney replied:
Please stop spamming me with email updates about this, The mute link does not work. Someone incorrectly mentioned me in some comment now I get 20-30 email spam messages a day from this thread. Someone remove me.

justinsb commented on test/integration/pods/pods_test.go:
```diff
@@ -155,7 +157,9 @@ func TestPodReadOnlyFilesystem(t *testing.T) {
 	ns := framework.CreateTestingNamespace("pod-readonly-root", s, t)
 	defer framework.DeleteTestingNamespace(ns, s, t)
 
-	client := client.NewOrDie(&restclient.Config{Host: s.URL, ContentConfig: restclient.ContentConfig{GroupVersion: testapi.Default.GroupVersion()}})
```
Please remove if more than just formatting

justinblayney replied:
Please stop spamming me with email updates about this, The mute link does not work. Someone incorrectly mentioned me in some comment now I get 20-30 email spam messages a day from this thread. Someone remove me.

justinsb replied:
Hi similarly-named justin :-)  I don't think anyone can do that for you - you have to do it yourself.  But if there is something we can do, please send us a link to the github instructions on how to do it.  I think it's just a limitation of github :-(

justinblayney replied:
I tried, I can't, the mute thread link does nothing. I think someone added me incorrectly thinking I was you a long time ago. I used to only get a msg once a month so I ignored it, now it qualifies as spam I get so many. I also develop, just not this

justinsb replied:
There is a "contact a human" button on the page: https://help.github.com/articles/unsubscribing-from-conversations/

My guess is that it is a github bug where you were mentioned previously and somehow that is overriding your mute every time.

(You can probably also mute the thread in gmail if you're using gmail)

justinblayney replied:
Hi there, i tried all that including contacting github, even now while tying this message in the right tool bar it asks me to subscribe "Subscribe, you're not receiving notifications from this thread". I was opening that whoever opened this project to begin with could remove me.

justinsb replied:
I'm sure we can find someone with the required privileges, if you can tell us what needs to be done!  What did github support say?

justinblayney commented:
Please stop spamming me with email updates about this, The mute link does not work. Someone incorrectly mentioned me in some comment now I get 20-30 email spam messages a day from this thread. Someone remove me.

justinsb commented on pkg/controller/job/jobcontroller_test.go:
```diff
@@ -220,7 +220,7 @@ func TestControllerSyncJob(t *testing.T) {
 
 	for name, tc := range testCases {
 		// job manager setup
-		clientset := clientset.NewForConfigOrDie(&restclient.Config{Host: "", ContentConfig: restclient.ContentConfig{GroupVersion: testapi.Default.GroupVersion()}})
```
My comment here seems to have been lost.  Are these `Host: ""` removals needed?

dshulyak replied:
:( i think thats because Hosts was changed everywhere and then i simply removed them instead of adding Host: "" back

dshulyak commented on pkg/client/restclient/request.go:
```diff
@@ -797,22 +798,31 @@ func (r *Request) request(fn func(*http.Request, *http.Response)) error {
-			return err
-		}
-		req.Header = r.headers
+		err := r.urlProvider.WrapWithRetry(
```
@justinsb What do you think? Is this better then to count urls?

justinsb replied:
Ah this is probably better - moves the decision making on retries into the url provider.

Is there a design for how we intend retries to work? (whether or not we implement them in this PR).  We should probably look at e.g. how GRPC does it.

Or maybe we should just adopt GRPC and be done with it :-)

krousey commented:
Reviewed 4 of 13 files at r12, 6 of 11 files at r13, 7 of 15 files at r14.
Review status: 21 of 24 files reviewed at latest revision, 54 unresolved discussions, some commit checks failed.

---

*[pkg/client/restclient/config.go, line 159 \[r14\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588#-KPo7qnZsX8U_X60PxgP:-KPo7qnZsX8U_X60PxgQ:boapbih) ([raw file](https://github.com/kubernetes/kubernetes/blob/ddda8403b359c1326cb6633893ddd386671f47c6/pkg/client/restclient/config.go#L159)):*
> ```Go
> }
> 
> // Returns Host and AlternateHosts as a single slice, if Host is already present
> ```

// Hosts returns Host and AlternateHosts...

---

*[pkg/client/restclient/config_test.go, line 97 \[r14\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588#-KPo8f0XHsfM4WAAMKsV:-KPo8f0XHsfM4WAAMKsW:boxuxhm) ([raw file](https://github.com/kubernetes/kubernetes/blob/ddda8403b359c1326cb6633893ddd386671f47c6/pkg/client/restclient/config_test.go#L97)):*
> ```Go
> 	for i, testCase := range testCases {
> 		if err := SetKubernetesDefaults(testCase.Config); err != nil {
> 			t.Errorf("%v: setting defaults failed for %#v: %v", i, testCase.Config, err)
> ```

%d is fine here and below.

---

*[pkg/client/restclient/urlprovider.go, line 32 \[r13\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588#-KP8Clqvt2DkeqMgMd7F-r13-32:-KPoR6X16CheTF8zNbkv:bxloxou) ([raw file](https://github.com/kubernetes/kubernetes/blob/12c65cf1a454d0595e0fc4fefb68c221c1f07eba/pkg/client/restclient/urlprovider.go#L32)):*
<details><summary><i>Previously, dshulyak (Dmitry Shulyak) wrote…</i></summary>
> maybe i shouldn't be adding this interface at all and just use RoundRobinProvider everywhere
</details>
I think a better interface would be to return a slice of all URLs and let them do whatever retry logic they want with that.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30588)*
<!-- Sent from Reviewable.io -->

justinsb commented on pkg/client/restclient/request.go:
```diff
@@ -687,17 +689,16 @@ func (r *Request) Watch() (watch.Interface, error) {
-	if req.baseURL != nil {
-		url = req.baseURL.Host
+	host := "none"
+	if u := req.urlProvider.Get(); u != nil {
```
What if URLProvider was updated in the meantime?

dshulyak replied:
changed to reuse URL on request object

dshulyak commented on pkg/client/restclient/versions.go:
```diff
@@ -46,10 +46,11 @@ func ServerAPIVersions(c *Config) (groupVersions []string, err error) {
 	if err != nil {
 		return nil, err
 	}
+	baseURL := baseURLs[0]
```
@justinsb somehow your comment was lost, this is executed during controller-manager startup. we are usually running api and controller-manager on the same node, configured to use local api server.

I guess i should change this as well, but it is not as critical as kube-proxy/kubelet parts. I hope it is not a problem if i will open issue for this problem and close it with small atomic pull request?

dshulyak commented:
@krousey @smarterclayton I verified this patch with kargo tool and 2 api servers. Is patch ok, or i should change something to get this merged?

smarterclayton commented on pkg/client/unversioned/clientcmd/client_config.go:
```diff
@@ -88,6 +88,21 @@ func (config *DirectClientConfig) RawConfig() (clientcmdapi.Config, error) {
 
+// cleanHost cleans URI, checks if any opaque data is present and
+// path is longer than 1 symbol.
+func cleanHost(host string) (string, error) {
```
@liggitt for security review of this change

smarterclayton commented on pkg/client/restclient/request.go:
```diff
@@ -712,26 +721,33 @@ func (r *Request) Stream() (io.ReadCloser, error) {
+	var req *http.Request
+	var resp *http.Response
+	err := r.urlProvider.WrapWithRetry(
+		func() error {
```
This block now looks identical across all the places it's called here.  Is that true?

If so please leave a TODO in the code that this needs to be refactored to be the same, and that req/resp should be returned values from the function, not closure parameters.

smarterclayton commented:
I think this needs an e2e test that actually uses it (if the goal is to get this in to 1.4).

liggitt commented on cmd/kube-proxy/app/options/options.go:
```diff
@@ -65,6 +66,7 @@ func NewProxyConfig() *ProxyServerConfig {
 func (s *ProxyServerConfig) AddFlags(fs *pflag.FlagSet) {
 	fs.Var(componentconfig.IPVar{Val: &s.BindAddress}, "bind-address", "The IP address for the proxy server to serve on (set to 0.0.0.0 for all interfaces)")
 	fs.StringVar(&s.Master, "master", s.Master, "The address of the Kubernetes API server (overrides any value in kubeconfig)")
+	fs.StringSliceVar(&s.APIServerList, "api-servers", []string{}, "List of Kubernetes API servers for publishing events, and reading pods and services. Comma separated list of hostnames or host:port pairs.")
```
why are we adding connection options outside of kubeconfig? I thought we were trying to unify the code path in the kubelet to get rid of the parallel options. If we want the ability to specify multiple hosts for a cluster, kubeconfig needs to be able to specify that

dshulyak replied:
i added api-servers to kube-proxy so that it will be uniform with kubelet, and ofcourse there is Servers options for kubeconfig

liggitt commented on pkg/client/unversioned/clientcmd/api/types.go:
```diff
@@ -62,6 +62,8 @@ type Cluster struct {
 	// Server is the address of the kubernetes cluster (https://hostname:port).
 	Server string `json:"server"`
+	// Servers is an array of addresses to the kubernetes apiservers with the format: (https://hostname:port)
+	Servers []string `json:"servers,omitempty"`
```
the names here are unclear if the host specified in `Server` should also be specified in `Servers`. Is the goal to make people specify a primary server in `Server` and fallbacks in `Servers`, or to specify the whole list in `Servers` and leave `Server` empty?

liggitt replied:
ah, missed the doc in v1/types.go

smarterclayton replied:
The former (alternate hosts are additional)

liggitt commented on pkg/client/restclient/config.go:
```diff
@@ -278,16 +298,29 @@ func InClusterConfig() (*Config, error) {
+		return true, nil
 	}
-	return baseURL.Scheme == "https"
+	return false, fmt.Errorf("All or none hosts must be TLS enabled %v", hosts)
```
`cannot mix https and non-https hosts`?

liggitt commented on cmd/kubelet/app/server.go:
```diff
@@ -553,7 +553,7 @@ func createClientConfig(s *options.KubeletServer) (*restclient.Config, error) {
 		if err != nil {
 			return nil, err
 		}
-		authConfig.Host = s.APIServerList[0]
```
this looks like it could leave `Host` empty. `AlternateHosts` is an odd field name if the primary `Host` field can be unset.

dshulyak replied:
i agree with you, in previous versions of the patch only Hosts were used (instead of AlternateHosts and Host) but couple of reviewers suggested to reduce scope of this change and make it additive

liggitt commented on pkg/client/restclient/urlprovider.go:
```diff
@@ -0,0 +1,84 @@
+	for range p.urls {
+		err = f()
+		if err != nil {
+			p.Next()
```
this assumes no other calls to Next() are made while this loop is running

dshulyak commented:
I will try to make kube-apiserver pod to listen on multiple host ports, change configuration and drop traffic to first one

mikedanese commented on pkg/client/restclient/urlprovider.go:
```diff
@@ -0,0 +1,84 @@
+}
+
+// A RoundRobinProvider allows to iterate over URLs, doesnt provide any health checking.
+type RoundRobinProvider struct {
```
I'm fairly certain we want session affinity for this client.

mikedanese replied:
The default kubernetes service has session affinity. We talked about this being preferable in a sig-ha meeting a while ago. For reference https://github.com/kubernetes/kubernetes/pull/23129#issuecomment-197990470

dshulyak replied:
is there any ongoing work to add session object to restclient?

dshulyak replied:
what if api servers are behind lb (haproxy) and one died in the middle of session? how we can preserve session affinity in such case?

i don't understand what is the proposal, add session object to restclient and re-work parts that need to hit same api server or you had completely different things in mind?

mikedanese replied:
Round robin is not an optimal strategy for clients connecting to the Kubernetes API Server. Clients should prefer to connect to the same API Server until it starts to fail repeatedly. Adding this capability to the client is most useful when the clients connect to the API directly. The kubernetes service in the default namespace implments this by using service SessionAffinity. You could implement this with HA proxy using "balance source" and "hash-type consistent". ELBs and GCPLBs support this as well. This was also the design of the original provider https://github.com/dshulyak/kubernetes/blob/c628b0f279a2581b3b59d43751289cf66740c65a/pkg/client/restclient/slightly_sticky_provider.go

This is definitely somehting that we want but I'll let someone else decide how important it is. cc @lavalamp

lavalamp replied:
> Clients should prefer to connect to the same API Server until it starts
to fail repeatedly

+1

On Thu, Sep 1, 2016 at 1:10 PM, Mike Danese <notifications@github.com>
wrote:

> In pkg/client/restclient/urlprovider.go
> <https://github.com/kubernetes/kubernetes/pull/30588#discussion_r77244373>
> :
>
> > +)
> > +
> > +// A URLProvider allows to use multiple URLs.
> > +// Also maintains state of currently working URL.
> > +type URLProvider interface {
> > +	// Get returns currently selected URL.
> > +	Get() *url.URL
> > +	// Next selects any other URL, it is upto implementation to provide any
> > +	// health checking/load balancing, it might be as simple as an iterator.
> > +	Next() *url.URL
> > +	// WrapWithRetry executes function until success or until all urls were visited.
> > +	WrapWithRetry(func() error) error
> > +}
> > +
> > +// A RoundRobinProvider allows to iterate over URLs, doesnt provide any health checking.
> > +type RoundRobinProvider struct {
>
> Round robin is not an optimal strategy for clients connecting to the
> Kubernetes API Server. Clients should prefer to connect to the same API
> Server until it starts to fail repeatedly. Adding this capability to the
> client is most useful when the clients connect to the API directly. The
> kubernetes service in the default namespace implments this by using service
> SessionAffinity. You could implement this with HA proxy using "balance
> source" and "hash-type consistent". ELBs and GCPLBs support this as well.
> This was also the design of the original provider
> https://github.com/dshulyak/kubernetes/blob/c628b0f279a2581b3b59d43751289c
> f66740c65a/pkg/client/restclient/slightly_sticky_provider.go
>
> This is definitely somehting that we want but I'll let someone else decide
> how important it is. cc @lavalamp <https://github.com/lavalamp>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30588/files/f80530c7f77e2fd272e6718b7d1754df0869d478#r77244373>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAngluqyHVih1A1_by4u3iggokMm_30Kks5qlzEagaJpZM4Jj5SF>
> .
>

dshulyak replied:
I thought that fail and retry would be preferred behavior. I will reuse previous design idea and will switch to next host only if errors burst will be high enough

mikedanese replied:
Pasting from a previous thread: The idea is that if we start to run controllers like scheduler and controller-manager onto the cluster in HA configurations, they should talk to the same API server if possible. Because each apiserver serves from a cache (unless the linearize read flag is passed explicitly) and each apiserver cache could be a different degree behind etcd. This could cause bugs in naive clients that don't handle this. We can avoid these bugs by adding clientIP session affinity.

We should do this but I'm not sure if we need it in the initial implementation.

smarterclayton replied:
I would prefer to only switch on error as well.  How complex is it to write the "backend preference" url provider and then cleanly integrate into rest client (does it require changes to the interface you are providing because we need to know more info)?

dshulyak replied:
1. In current implementation switch will happen only on error, e.g. there is no internal load balancer, and we will stick to the first url as long as it will work.
Sure, Next is public, but it is used only for testing.

2. There is 2 differences with https://github.com/dshulyak/kubernetes/blob/c628b0f279a2581b3b59d43751289cf66740c65a/pkg/client/restclient/slightly_sticky_provider.go, in slightly sticky provider implementation:

   - URL selected based on time
   - switch happens not on **first** error, but certain burst of errors can be tolerated

    Why those things will provide stronger session affinity guarantees?

3. @smarterclayton you are probably right about interface. If we will want to provide such configuration as errors qps and errors burst, we will need to change interface. But it can be done in following patches

mikedanese replied:
> Why those things will provide stronger session affinity guarantees?


> URL selected based on time

The url is selected based on the output of random number generator. The RNG is seeded by the current time. We don't want all clients to start with using the first  url in the list. That is unlikely to cause even balancing. As far as the url successor funciton, I think random is preferable to round robin for similar reasons.

> switch happens not on first error, but certain burst of errors can be tolerated

To allow spurious/transient failure to pass without re-balancing clients i.e. to allow a configurable level of forgiveness.

dshulyak commented:
I wasn't able to write a good e2e test for this change, which will pass on gce provider. Mainly because of current deployment model. It should work but it is quite ugly, thats why i decided to extract proposed e2e test and continue discussion in separate PR (https://github.com/kubernetes/kubernetes/pull/32380). 

@smarterclayton @krousey are you fine with this?

ssboisen commented:
Will this make it to 1.4?

dshulyak commented:
@ssboisen it definitely won't be in 1.4.0 release. I will do some additional work on testing and hope to land it in the beginning of 1.5

Also i will discuss option to cherry-pick it into stable-1.4 branch, and release as a part of 1.4.1

dshulyak commented:
@k8s-bot gce e2e test this

dshulyak commented:
@k8s-bot unit test this

dshulyak commented:
@smarterclayton @krousey Can this change be accepted as it is implemented right now?

I added couple of integration test cases to verify url provider behavior.
As for the session affinity, i would prefer to address those concerns in new PR, if thats possible.

Also I will work on advanced e2e test cases once we will have multi api-server deployment on CI. I can try to reuse #32860, but it stills requires updating kubelet configuration dynamically, which is possible but a bit problematic.

PS: unit test failures seems to me as unrelated, i will figure where is the problem

smarterclayton commented:
I'll defer to @mikedanese since this was his vision

mikedanese commented:
@dshulyak Please open issues for open work items. I'm ok with an initial implementation to build on, although I have not looked deeply at the current code.

dshulyak commented:
@k8s-bot unit test this

dshulyak commented:
@smarterclayton @mikedanese I opened issues for all related work items. Please consider merging or maybe final review

justinsb commented:
Is there a description/design doc for how this feature would be used?  I think this may be better handled through DNS.

dshulyak commented:
@justinsb There is not much to describe. There is a flag for kubelet `-api-servers`, but clients were using only first host from a list. I fixed it, also hosts can be provided by using kubeconfig. I can provide documentation once it will get merged.

I don't think that we want to rely on DNS server configured and maintained by someone else, for instance we will have to change dns configuration each time we will add/remove api server node.
As far as I can tell a lot of people want to have it out of the box, why not?

liggitt commented:
> There is not much to describe. There is a flag for kubelet -api-servers, but clients were using only first host from a list. I fixed it, also hosts can be provided by using kubeconfig. I can provide documentation once it will get merged.

The plumbing doesn't need description, but the algorithm for how clients detect the need to fallback to the next server in the list, and how fallback behaves certainly needs discussion/description (is the request that failed and triggered fallback retried? how is the next server selected? does the client partition the list into "healthy" and "unhealthy" servers? etc.)

justinsb commented:
I think you're spot on @liggitt - and it's because of that complexity that I think we want to be very sure that we want to do this.  I realize the flag previously implied that you could pass multiple apiservers, but does anyone actually intend to use this?

I think the flag was originally designed before we realized that golang handled multiple A records pretty well (or golang's handling wasn't so good at the time).  The reason _not_ to merge code is that it is complex; we'll have to maintain this going forwards.  So I just want to know if there's anyone that would actually use this, and what for!

It seems to me that most configurations will actually prefer a load balancer or DNS.  The problem with multiple IP addresses is that we then have to figure out how to update that set of IP addresses, in particular in the case when they _all_ change.  Every production configuration I have seen has therefore used IP redirection, a load balancer or DNS.

>  for instance we will have to change dns configuration each time we will add/remove api server node.

Isn't it much easier to update DNS/your load balancer in one place than it is to update every kubelet & kube-proxy config, and every kubeconfig file on every client's machine?

Bregor commented:
@justinsb but what if one of API servers dying by night and nobody knows it till morning (yes I know it's monitoring problem but anyway)? DNS still will be pointing to dead server and 50% of requests will fail.
Am I wrong?

smarterclayton commented:
Part of the goal here is to support multiple endpoints (which could be multiple DNS as well) so that the client can "cut" over from one group to another.  The canonical example would be:

1. Admin sets up a "bootstrap DNS" that points to a VIP or can be updated
2. The bootstrap DNS resolves to a known apiserver 
3. The clients are configured to talk to bootstrap DNS, but to also talk to a set of apiservers identified by cluster IPs that may be changed on the fly (a list of endpoints)
4. If all of the cluster IPs go down, the client can still talk to the bootstrap API server to get the other endpoints

So this is mostly orthogonal to DNS vs not DNS - just ensuring there is > 1 path from a given client back to an API server, and that the client will eventually try all API servers in its list in order to find a working path.

justinsb commented:
@Bregor based on the golang code, and experimentally, the client will try all the IP addresses that are resolved: https://github.com/kubernetes/kubernetes/issues/18174#issuecomment-232850783

So clients will work, as long as at least one IP is still functional.

justinsb commented:
@smarterclayton is that something you're doing today in OpenShift, or is that the proposed sig-cluster-lifecycle work?

dshulyak commented:
> So I just want to know if there's anyone that would actually use this, and what for!

We will actually use it in our installer as soon as this change will be released.

> Isn't it much easier to update DNS/your load balancer in one place than it is to update every kubelet & kube-proxy config, and every kubeconfig file on every client's machine?

If we don't own DNS server it is not easier. It will require us to integrate with customer orchestration.

Also i am not sure that with DNS server we will be able to support all logic discussed in this thread https://github.com/kubernetes/kubernetes/pull/30588/files#r77038792

dshulyak commented:
@k8s-bot gce e2e test this

dshulyak commented:
@smarterclayton @krousey Please tell me what is currently blocks this change.

If your primary concern is quality of this change, i can assure you that @kubernetes/mirantis will start using it in downstream CI immediately and either me or another team member will commit to fix/troubleshoot bugs in this area as soon as they will be discovered.

Also we will for sure prepare all required documentation for 1.5 release that is related to this change.

krousey commented on pkg/client/restclient/request.go:
```diff
@@ -786,22 +804,31 @@ func (r *Request) request(fn func(*http.Request, *http.Response)) error {
-		}
-		req.Header = r.headers
+		// TODO dedup multiple usage of WrapWithRetry in Stream/Watch/request methods
+		err := r.urlProvider.WrapWithRetry(
```
I think I mentioned this a while ago, but I really don't like this pattern. This inverts the control of retries and hands it to the `urlProvider`. It doesn't respect the `maxRetries` defined above. And it's also not-integrated with the server-side rate-limiting (line 849 below). It looks to me that a RATE_LIMIT_EXCEEDED response from one server will cause the client to sleep, and then move on to a server that wasn't rate limiting.

I guess that is to say `urlProvider` should really only have one job: provide URLs. I would much rather see an implementation where a `urlProvider` provides a snapshot slice of URLs that are currently considered valid, and let this client do retry strategies/load balancing how it sees fit.

bogdando replied:
Looking to this endless log of comments for endless review, just wondering if accepting the *basic* implementation for multi-hosts access and *then* improving it with rate limiting and other things would make sense? The thing is very nice to have. F.e. etcd cluster can already be accessed in multi-hosts fashoin by *all* cluster components including network plugins. This allowed us (in Kargo, https://github.com/kubernetes-incubator/kargo/pull/591) to drop etcd-proxy from HA layout and just live with multi-access mode. The same is expected by ops with Apiserver.

krousey reviewed (changes requested):
Sorry this sat for so long. I assumed clayton was looking at it.

luxas commented:
Ping @dshulyak @smarterclayton @krousey 

I guess we want to proceed with this, and get it into v1.6
This will be crucial before we can implement HA "for real"

luxas commented:
@smarterclayton @justinsb @dshulyak @mikedanese @krousey ping
Can we get back to this now, please?

And a question, what's actually blocking us from being able to parse the CLI flag `--master` which is a string into a `[]string` internally by splitting the `--master` flag on commas?
e.g. `--masters https://foo-ip:6443,https://dnsname1` would just become a `[]string{"https://foo-ip:6443", "https://dnsname1"}` or `{Host: "https://foo-ip:6443", AlternateHosts: []string{"https://dnsname1"}` if you like.

I know it may seem a little bit hacky, but why not actually? It would solve most of the problems I guess.
I know it wouldn't fix the KubeConfig file, but anyway.
WDYT?

dshulyak commented:
I will rework this change as @krousey suggested

dshulyak commented:
@k8s-bot bazel test this

dshulyak commented:
@krousey Can you please review it again? I reworked change as you suggested, also simplified it. In current version there will be no automatic retries and restclient will always try to use first valid url in list. URL will be invalidated in case of error and client will try to use another URL. 

I plan to add rate limiters to URLContainer in subsequent requests to provide some stickiness.

@caesarxuchao @lavalamp I noticed that client-go is used in kubelet. Where I can read about release process for client-go and it relation to pkg/client? Do I need to propose changes for client-go directly?

luxas commented:
@mikedanese PTAL ^

dshulyak committed:
Allow to use multiple hosts in kubernetes rest client

dshulyak committed:
Update bazel

dshulyak commented:
@k8s-bot gce etcd3 e2e test this

ant31 commented:
How can we progress on this issue?

dshulyak commented:
Closing current PR in favor of https://github.com/kubernetes/kubernetes/pull/40674
